File: php/Day10.php
<?php
    require_once('Utils.php');
    
    function Day10_Part1() {
        $opens = array("(", "[", "{", "<");
        $closes = array(")", "]", "}", ">");
        $incorrects = array(")" => 0, "]" => 0, "}" => 0, ">" => 0);

        foreach (read_file_lines("../data/day10.txt") as $line) {
            $pairs = array();
            foreach (str_split($line) as $char) {
                if (in_array($char, $opens))
                    array_push($pairs, $char);
                else if (array_search($pairs[count($pairs) - 1], $opens) == array_search($char, $closes))
                    array_pop($pairs);
                else {
                    $incorrects[$char]++;
                    break;
                }
            }
        }

        $result = $incorrects[")"] * 3 + $incorrects["]"] * 57 + $incorrects["}"] * 1197 + $incorrects[">"] * 25137;

        echo "Day: 10 | Part: 1 | Result: " . $result . "<br>";
    }

    function Day10_Part2() {
        $opens = array("(", "[", "{", "<");
        $closes = array(")", "]", "}", ">");
        $scores = array();

        foreach (read_file_lines("../data/day10.txt") as $line) {
            $pairs = array();
            $score = 0;
            $corrupted = false;

            foreach (str_split($line) as $char) {
                if (in_array($char, $opens))
                    array_push($pairs, $char);
                else if (array_search($pairs[count($pairs) - 1], $opens) == array_search($char, $closes))
                    array_pop($pairs);
                else {
                    $corrupted = true;
                    break;
                }
            }

            if ($corrupted == false)
            {
                while (count($pairs) > 0)
                    $score = $score * 5 + array_search(array_pop($pairs), $opens) + 1;
                array_push($scores, $score);
            }
        }

        sort($scores);
        $result = $scores[count($scores) / 2];

        echo "Day: 10 | Part: 1 | Result: " . $result . "<br>";
    }

    Day10_Part1();
    Day10_Part2();

File: go/Day10.go
package main

import (
	"aoc2021/Utils"
	"fmt"
	"sort"
)

func Day10_Part1() {
	var opens = []string{"(", "[", "{", "<"}
	var closes = []string{")", "]", "}", ">"}
	incorects := map[string]int{")": 0, "]": 0, "}": 0, ">": 0}

	lines := Utils.ReadFile("../data/day10.txt")

	for line_id := range lines {
		pairs := make([]string, 0)
		for char_id := range lines[line_id] {
			if Utils.InArrayString(opens, string(lines[line_id][char_id])) {
				pairs = append(pairs, string(lines[line_id][char_id]))
			} else if Utils.IndexOf(opens, pairs[len(pairs)-1]) == Utils.IndexOf(closes, string(lines[line_id][char_id])) {
				pairs = append(pairs[:(len(pairs)-1)], pairs[len(pairs):]...)
			} else {
				incorects[string(lines[line_id][char_id])]++
				break
			}
		}
	}

	var result int = incorects[")"]*3 + incorects["]"]*57 + incorects["}"]*1197 + incorects[">"]*25137

	fmt.Printf("Day: 10 | Part: 1 | Result: %d\n", result)
}

func Day10_Part2() {
	var opens = []string{"(", "[", "{", "<"}
	var closes = []string{")", "]", "}", ">"}
	costs := make([]int, 0)

	lines := Utils.ReadFile("../data/day10.txt")

	for line_id := range lines {
		pairs := make([]string, 0)
		var cost int = 0
		var corrupted bool = false

		for char_id := range lines[line_id] {
			if Utils.InArrayString(opens, string(lines[line_id][char_id])) {
				pairs = append(pairs, string(lines[line_id][char_id]))
			} else if Utils.IndexOf(opens, pairs[len(pairs)-1]) == Utils.IndexOf(closes, string(lines[line_id][char_id])) {
				pairs = Utils.RemoveAt(pairs, len(pairs)-1)
			} else {
				corrupted = true
				break
			}
		}

		if !corrupted {
			for len(pairs) > 0 {
				cost = cost*5 + Utils.IndexOf(opens, pairs[len(pairs)-1]) + 1
				pairs = Utils.RemoveAt(pairs, len(pairs)-1)
			}
			costs = append(costs, cost)
		}
	}

	sort.Ints(costs[:])
	var result int = costs[len(costs)/2]

	fmt.Printf("Day: 10 | Part: 2 | Result: %d\n", result)
}
